### PR TITLE
Gpu features detection

### PR DESCRIPTION
**Pull request by edgarriba**
Enable to compute ORB features using a GPU in case that OpenCV is built with CUDA

### CODE REVIEW

File: bin/detect_features [hunk 1]
```diff
@@ -33,8 +33,9 @@ def detect(image):
         f_pre = f_sorted[-preemptive_max:]
         data.save_features(image, p_sorted, f_sorted, c_sorted)
         data.save_preemptive_features(image, p_pre, f_pre)
-        index = features.build_flann_index(f_sorted, data.config)
-        data.save_feature_index(image, index)
+        # got (seg fault)
+        #index = features.build_flann_index(f_sorted, data.config)
```
Comment (edgarriba) [inline bin/detect_features]: it takes too much time here and finally got seg fault

**Comment by edgarriba** [comment]
@paulinus Finally getting successful results in features detection and matching
figure_1
figure_2
reconstruction_berlin_orb

On the other hand, reconstruction seems to succeed by looking at the 3D visualization but in the end Ceres is giving a FAILURE report.

eriba@CVC224:~/software/OpenSfM$ ./bin/run_all data/berlin
Extracting focal lengths for image 02.jpg
Extracting focal lengths for image 03.jpg
Extracting focal lengths for image 01.jpg
Device 0:  "GeForce GTX 970"  4096Mb, sm_52, Driver/Runtime ver.7.50/7.50
Loading ORB-GPU ...
Loading ORB-GPU ... [OK]
INFO:__main__:Extracting ROOT_ORB features for image 02.jpg
DEBUG:opensfm.features:Computing ORB-GPU
DEBUG:opensfm.features:Found 48556 points in 0.024836063385s
INFO:__main__:Extracting ROOT_ORB features for image 03.jpg
DEBUG:opensfm.features:Computing ORB-GPU
DEBUG:opensfm.features:Found 43976 points in 0.0196778774261s
INFO:__main__:Extracting ROOT_ORB features for image 01.jpg
DEBUG:opensfm.features:Computing ORB-GPU
DEBUG:opensfm.features:Found 49741 points in 0.0249769687653s
Matching 3 image pairs
Matching 02.jpg  -  1 / 3
Matching 01.jpg  -  2 / 3
Matching 03.jpg  -  3 / 3
len(matches) 614
Robust matching time : 0.121917009354s
Full matching 559 / 614, time: 18.3632371426s
len(matches) 1793
Robust matching time : 0.0942587852478s
Full matching 1767 / 1793, time: 20.9849448204s
len(matches) 164
Robust matching time : 0.000285863876343s
Full matching 152 / 164, time: 16.7249338627s
2015-11-09 12:54:45,662 reading features
2015-11-09 12:54:45,718 Merging features onto tracks
2015-11-09 12:54:45,730 Good tracks: 2378
images 3
nonfisheye images 3
Initial reconstruction with 02.jpg and 01.jpg
Number of common tracks 1780
Number of inliers 1780
Number of reconstructed 3D points :1776
Found initialize good pair 02.jpg and 01.jpg
Ceres Solver Report: Iterations: -2, Initial cost: -1.000000e+00, Final cost: -1.000000e+00, Termination: FAILURE
setup/run/teardown 0.00577688217163/0.00767612457275/0.00237393379211
Resection 03.jpg inliers: 66 / 80
 -------------------------------------------------------
Adding 03.jpg to the reconstruction
Ceres Solver Report: Iterations: -2, Initial cost: -1.000000e+00, Final cost: -1.000000e+00, Termination: FAILURE
setup/run/teardown 0.00762581825256/0.0101351737976/0.00371599197388
Remove 9 outliers
Ceres Solver Report: Iterations: -2, Initial cost: -1.000000e+00, Final cost: -1.000000e+00, Termination: FAILURE
setup/run/teardown 0.0074360370636/0.0103259086609/0.00348210334778
Reprojection Error: 0.00034689033676
Painting the reconstruction from 3 cameras
Done.
Reconstruction 0 : 3 images , 2223 points
1 partial reconstructions in total.

File: opensfm/src/bfm_gpu.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  // Perform Brute-Force matching
+  bfm_->knnMatch(d_query_desc, d_train_desc, matches_cpu, k);
+
+  //std::cout << matches_cpu.size() << std::endl;
+
+  // Convert to numpy.
+  bp::list retn;
+  for (size_t i = 0; i < matches_cpu.size(); ++i)
+    retn.append(std_vector_to_py_list(matches_cpu[i]));
+
```
Comment (edgarriba) [inline opensfm/src/bfm_gpu.h]: @paulinus Here got the following error when trying to return a list of lists of DMatch. Any idea?

eriba@CVC224:~/software/OpenSfM$ ./bin/match_features data/berlin
Matching 3 image pairs
Matching 02.jpg  -  1 / 3
Loading BFMatcherGpu ...
Loading BFMatcherGpu ... [OK]
Traceback (most recent call last):
  File "./bin/match_features", line 162, in <module>
    match(arg)
  File "./bin/match_features", line 96, in match
    matches = matching.match_symmetric(f1, i1, f2, i2, data.config)
  File "/home/eriba/software/OpenSfM/opensfm/matching.py", line 31, in match_symmetric
    matches_ij = [(a,b) for a,b in match_lowe_bf(fi, fj, config)]
  File "/home/eriba/software/OpenSfM/opensfm/matching.py", line 65, in match_lowe_bf
    matches = matcher.knnMatch(f1, f2, 2)
TypeError: No to_python (by-value) converter found for C++ type: cv::DMatch

**Comment by russkel** [comment]
Numba has GPU support, is this worth looking into? A pure python implementation would be easier to maintain.

**Pull request closed without merging**